tests/unit/lms/views/api/canvas/sync_test.py

```diff
@@ -1,140 +1,148 @@
+    context_id = request_json["course"]["context_id"]
+
+    def _groups_for(sections):
+        return [
+            HGroup.from_lti_parts(
+                name=section.get("name", f"Section {section['id']}"),
+                tool_consumer_instance_guid=tool_guid,
+                context_id=context_id,
+                section_id=section["id"],
+                type_="section_group",
+            )
+            for section in sections
+        ]
+
+    def assert_return_values(groupids, synced_sections, returned_sections):
+        lti_h_service.sync.assert_called_once_with(
+            _groups_for(synced_sections), request_json["group_info"]
+        )
+
+        assert groupids == [
+            group.groupid(TEST_SETTINGS["h_authority"])
+            for group in _groups_for(returned_sections)
+        ]
+
+    return assert_return_values
 
 
 @pytest.fixture
-def canvas_api_client(canvas_api_client):
-    # The course has three sections named "Section 1", "Section 2" and "Section
-    # 3" (with IDs 1, 2 and 3).
-    canvas_api_client.course_sections.return_value = [
-        {"id": i, "name": f"Section {i}",} for i in range(1, 4)
-    ]
+def is_instructor(pyramid_request):
+    pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Instructor")
 
-    # The learner is only a member of section 2.
-    canvas_api_client.authenticated_users_sections.return_value = [
-        {"id": 2, "name": "Section 2"}
-    ]
 
-    # users_sections() returns id's only, not names.
-    canvas_api_client.users_sections.return_value = [{"id": 2}]
+@pytest.fixture
+def is_learner(pyramid_request):
+    pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Learner")
+
+
+@pytest.fixture
+def is_speedgrader(request_json):
+    request_json["learner"] = {"canvas_user_id": "user_id"}
+
+
+@pytest.fixture
+def sections():
+    class Sections:
+        # The course has three sections named "Section 1", "Section 2" and "Section
+        # 3" (with IDs 1, 2 and 3).
+        course = [{"id": i, "name": f"Section {i}"} for i in range(1, 4)]
+
+        # users_sections() returns id's only, not names.
+        authenticated_user = [{"id": 2, "name": "Section 2"}]
+
+        # The learner is only a member of section 2.
+        user = [{"id": section["id"]} for section in authenticated_user]
+
+    return Sections()
+
+
+@pytest.fixture
+def canvas_api_client(canvas_api_client, sections):
+    canvas_api_client.course_sections.return_value = sections.course
+    canvas_api_client.authenticated_users_sections.return_value = (
+        sections.authenticated_user
+    )
+    canvas_api_client.users_sections.return_value = sections.user
 
     return canvas_api_client
 
 
 @pytest.fixture
-def pyramid_request(pyramid_request):
-    pyramid_request.json = {
+def pyramid_request(pyramid_request, request_json):
+    pyramid_request.json = request_json
+    return pyramid_request
+
+
+@pytest.fixture
+def request_json():
```

Comment:
Similar to splitting the sections data out of `canvas_api_client` the request JSON body has been separated out to make it a bit easier to access.